### PR TITLE
Release 0.4.0 — observability operators (Tap/Log/Throttle) + test/supply-chain hardening

### LINKED ISSUE #76: [Maintenance] testing: Continuous fuzz running (CI-time hours) via FsCheck or CsCheck
## Scope

Per-PR property tests run in seconds (deterministic seeds, ~50 cases per property) — that's the short version of fuzzing. The deferred version is **continuous fuzz**: hundreds of thousands of randomized cases per property running for CI-time hours on a scheduled workflow, surfacing edge cases the short version misses.

## Acceptance criteria

- A new `tests/<Repo>.Tests.Fuzz/` project (or `[Trait("Category", "Fuzz")]` opt-in tests in the existing test project) uses FsCheck or CsCheck.
- Properties cover the public API surface: every public method that has a meaningful invariant (e.g. round-trip, idempotency, equivalence with a reference implementation).
- A new `.github/workflows/fuzz.yaml` runs on a weekly cron + `workflow_dispatch`. Generous case-count target per property (e.g. 100,000 cases × 5-minute time budget).
- Failing inputs saved to a fuzz-replay folder (committed back, or attached as a workflow artifact) so the next short-version run includes them as deterministic regressions.
- Findings auto-file an issue (`gh issue create` in the workflow) when a property fails.

## Notes

- Tier-1 follow-up — out of scope for the short-pass property tests in the unit suite.
- FsCheck vs CsCheck choice is open: FsCheck is more mature; CsCheck has better shrinking and is .NET-native idiomatic.
- Originated as part of the IAsyncEnumerable-Extensions thorough-review pass; tracked fleet-wide.
- Board: https://github.com/users/Chris-Wolfgang/projects/6

### LINKED ISSUE #77: [Maintenance] testing: Production-traffic shadow testing via sample consumer workloads
## Scope

Unit + property tests prove behavioural contract under *synthetic* input. They do not catch performance regressions or pathological allocation patterns that only surface under realistic production traffic shapes — wildly variable input sizes, bursty call rates, mixed sync/async sources, retry/backoff loops, concurrent callers.

The deferred work is to record (or synthesize) realistic consumer workloads and replay them against new library versions before release, comparing latency / allocation / GC against a golden baseline.

## Acceptance criteria

- One or more sample consumers under `samples/` exercise each public method in a realistic scenario appropriate to this library's domain (e.g. SQL paging, HTTP paging, file streaming, concurrent enumeration — depending on what the library actually does).
- A shadow-test workflow (`shadow.yaml`) runs the samples nightly against the latest commit and a baseline release, captures BDN-style results + GC stats, and fails if a metric regresses beyond a threshold (e.g. > 20% latency, > 50% allocations).
- Regressions vs baseline are posted as an issue; alerting thresholds are configurable per metric.

## Notes

- Tier-1 follow-up. Distinct from the P2 benchmark gh-pages chart issue (which runs curated BDN on every push to main and graphs the trend); shadow-testing runs realistic consumer scenarios.
- Sample consumers also double as the "real usage" documentation that synthetic XML-doc examples can't provide.
- Originated as part of the IAsyncEnumerable-Extensions thorough-review pass; tracked fleet-wide.
- Board: https://github.com/users/Chris-Wolfgang/projects/6

### LINKED ISSUE #78: [Maintenance] security: Security-grade static analysis (SAST) beyond CodeQL
## Scope

This repo already runs CodeQL (`codeql.yaml`, security-extended query pack per the S1 canonical). That covers the OSS-grade SAST baseline.

The deferred work is the **paid-tier SAST tools** that catch issue classes CodeQL misses or surfaces with low confidence: Snyk Code (semantic taint), Veracode SAST, Fortify SCA, SonarQube Cloud, Checkmarx One. Each has a quality-vs-cost trade-off and a different scanner.

## Acceptance criteria

Choose one (or more) and wire into the canonical `pr.yaml` pipeline:

- Account + license obtained for the chosen tool(s).
- A new workflow (or step in `pr.yaml`) runs the scanner on every PR and reports findings as PR annotations.
- A baseline scan of the current `main` is committed so per-PR scans report only *new* findings, not the inherited baseline.
- Scanner findings are triaged (false-positive / fixed / accepted-risk) and the result is tracked in a `SECURITY.md` audit log.

## Notes

- Tier-1 follow-up. Strictly *additive* to CodeQL — the goal is overlapping signals, not replacing the existing scanner.
- Cost is the gating factor: Snyk Code has a free OSS tier (worth trying first); the enterprise tools are typically $$$ per developer per month.
- This is the canonical follow-up to flag during a security audit; the implementation tier (which tool, what budget) is a project-level decision, not a per-repo one.
- Originated as part of the IAsyncEnumerable-Extensions thorough-review pass; tracked fleet-wide.
- Board: https://github.com/users/Chris-Wolfgang/projects/6

### LINKED ISSUE #83: [Maintenance] API: Binary compatibility / ABI diff between releases (Microsoft.DotNet.ApiCompat)
## Scope

A1 (PublicApiAnalyzers, #199 family) catches added/removed signatures via `PublicAPI.Shipped.txt` diffs. It does NOT catch behavioural breaks like default-value changes, nullability annotation flips, or binary-layout shifts that consumers' compiled assemblies trip over at runtime.

The deferred work is to add **Microsoft.DotNet.ApiCompat** (or equivalent) as a release gate: every release validates ABI compatibility against the previous published version on NuGet.

## Acceptance criteria

- `release.yaml` gains an `api-compat` step that downloads the previous NuGet version and runs `ApiCompat <current>.dll <previous>.dll`.
- The step fails the release if a non-major version bump introduces an ABI break (per SemVer: PATCH/MINOR forbid ABI breaks; MAJOR allows them).
- The set of intentional breaks (in a MAJOR release) is recorded in a `compat-suppressions.txt` so they don't recur silently.

## Notes

- Tier-1 follow-up. Complementary to A1; strictly additive.
- Single biggest correctness signal for a NuGet library.
- Originated as part of the IAsyncEnumerable-Extensions thorough-review pass; tracked fleet-wide.
- Board: https://github.com/users/Chris-Wolfgang/projects/6

### LINKED ISSUE #84: [Maintenance] testing: Concurrency / race-condition stress testing (Coyote, soak runs)
## Scope

Per-PR tests run one thread. Production usage of an async library hits real schedule interleavings: concurrent consumers of the same `IAsyncEnumerable`, racing `DisposeAsync` calls, cancellation arriving in the middle of an `await` continuation that lands on a different worker.

The deferred work is **systematic schedule exploration** via Microsoft's [Coyote](https://github.com/microsoft/coyote) — a model checker for async/concurrent C#. Coyote takes existing tests and replays each one under thousands of randomly-explored interleavings, surfacing the race/deadlock paths that ran a million times in CI doesn't find.

Optionally also: 24-hour soak tests that hammer the public API and check for handle leaks, thread leaks, memory leaks via `dotnet-counters`.

## Acceptance criteria

- A `tests/<Repo>.Tests.Concurrency/` project (or `[Trait("Category", "Concurrency")]` opt-in) instruments key public APIs with Coyote.
- A scheduled workflow (weekly) runs Coyote with a generous schedule budget (e.g. 10K iterations / property).
- Soak workflow (monthly) runs the API in a loop for 24h on a self-hosted runner, captures `dotnet-counters` snapshots, fails on regression.

## Notes

- Tier-1 follow-up. The 4 concurrent-disposal tests in `DeepBehaviorTests` (PR #227 in IAsyncEnumerable-Extensions) are the static version; this is the systematic version.
- Originated as part of the IAsyncEnumerable-Extensions thorough-review pass; tracked fleet-wide.
- Board: https://github.com/users/Chris-Wolfgang/projects/6

### LINKED ISSUE #85: [Maintenance] security: Supply chain hardening — SBOM, SLSA attestation, package signing
## Scope

CodeQL (S1) covers SAST on the repo's own source. The deferred work is **supply-chain hardening** — proving that what consumers download from NuGet was actually built from this repo, with no malicious injection at any link:

- Generate an SBOM (CycloneDX or SPDX) at release time and attach to the GitHub Release
- Generate SLSA provenance attestation (level 2 or 3) via `actions/attest-build-provenance`
- Sign the NuGet package with a verifiable certificate (NuGet trusted signers feature, or Sigstore)
- Verify the SBOM + provenance + signature in a downstream consumer to prove the chain works end-to-end

## Acceptance criteria

- `release.yaml` generates SBOM + SLSA attestation; both attached to the GitHub Release.
- NuGet packages are signed; the signature is verifiable via `nuget verify`.
- A documented procedure (in `SECURITY.md`) explains how consumers verify the chain.

## Notes

- Tier-1 follow-up. Required by enterprise procurement and US EO 14028 / similar regulations.
- Strictly additive to CodeQL S1 + the SAST issue separately filed.
- Originated as part of the IAsyncEnumerable-Extensions thorough-review pass; tracked fleet-wide.
- Board: https://github.com/users/Chris-Wolfgang/projects/6

### LINKED ISSUE #86: [Maintenance] testing: Cross-platform / multi-arch differential (Linux/macOS/Windows × x64/ARM64)
## Scope

`pr.yaml` Stages 1/2/3 run the test suite on Linux, Windows, macOS — but the outputs are not *diffed* against each other. A library can pass on all three while behaving subtly differently on one (line-ending handling, culture-sensitive sorts, file-system case sensitivity, timing-dependent assertions).

ARM64 (Apple Silicon, AWS Graviton) is untested entirely. .NET supports it; the library hasn't been verified there.

## Acceptance criteria

- A `cross-platform-differential.yaml` workflow runs the same test set on linux-x64, linux-arm64, macos-arm64, windows-x64 and diffs their textual outputs (test names + assertion messages + any reported numbers).
- Any divergence (other than tagged platform-specific tests) fails the workflow.
- The matrix includes at least one ARM64 runner (linux-arm64 self-hosted or `ubuntu-24.04-arm`).

## Notes

- Tier-1 follow-up. Distinct from per-PR Stages 2/3 which just verify it passes on each OS — this verifies it passes the SAME WAY on each.
- Originated as part of the IAsyncEnumerable-Extensions thorough-review pass; tracked fleet-wide.
- Board: https://github.com/users/Chris-Wolfgang/projects/6

### LINKED ISSUE #87: [Maintenance] testing: Snapshot / approval testing for stable outputs (Verify framework)
## Scope

Where the library produces structured output (formatted strings, JSON, XML, CSV), snapshot testing via the [Verify](https://github.com/VerifyTests/Verify) framework catches accidental format drift that unit tests targeted assertions miss. The first run records the output as a `.verified.txt` file; subsequent runs diff against it.

## Acceptance criteria

- Verify package referenced in the test project.
- Snapshot tests cover every public method whose output is text-shaped, complex-shaped (records, anonymous objects), or has interesting edge-case formatting (culture-sensitive, timezone-sensitive, etc.).
- Snapshot files committed under `tests/.../Snapshots/`.
- CI verifies snapshots match; `.received.txt` files committed during local iteration are gitignored.

## Notes

- Tier-1 follow-up. Only meaningfully valuable where the library produces inspectable output; pure-data utility libraries can close as not-applicable.
- Originated as part of the IAsyncEnumerable-Extensions thorough-review pass; tracked fleet-wide.
- Board: https://github.com/users/Chris-Wolfgang/projects/6

### LINKED ISSUE #89: [Maintenance] performance: Long-running GC / allocation profiling (PerfView, dotnet-counters)
## Scope

BDN throughput benchmarks (P1) measure single-method micro-perf. The deferred work is **sustained-load GC / allocation profiling** via PerfView (or `dotnet-counters` + `dotnet-trace`) — characterise gen0/1/2 promotion rates, LOH pressure, finalizer queue depth, ThreadPool queue depth — the metrics that only matter under long-running server / ETL workloads.

## Acceptance criteria

- A scheduled workflow runs the library against a high-volume realistic workload for ≥ 10 minutes per scenario, captures ETW traces.
- Reports surface the top allocation sites, top object types in heap, gen2 / LOH frequency.
- Regressions vs a baseline release fail the workflow.

## Notes

- Repo-specific Tier-1 follow-up — only filed against ETL family + In-memory-Logger (the libraries actually used in long-running server contexts). Pure utility libraries get a pass.
- Originated as part of the IAsyncEnumerable-Extensions thorough-review pass.
- Board: https://github.com/users/Chris-Wolfgang/projects/6

### LINKED ISSUE #90: [Maintenance] testing: Trim / AOT compatibility verification (PublishAot smoke)
## Scope

.NET 8+ Native AOT (`<PublishAot>true</PublishAot>`) and trimming (`<PublishTrimmed>true</PublishTrimmed>`) are real consumer requirements. Our libraries advertise `net8.0` / `net10.0` targets but no one has verified they survive the trimmer — reflection on type names the trimmer can't see, dynamic generic instantiation, runtime code generation all break silently under AOT.

## Acceptance criteria

- A new workflow / test step publishes a tiny console-app consumer with `<PublishAot>true</PublishAot>` + `<PublishTrimmed>true</PublishTrimmed>` referencing the library NuGet (or `ProjectReference`).
- The consumer calls every public method at least once and the published binary runs without `MissingMethodException` / `NotSupportedException` / silent no-ops.
- Any required trim-warnings suppressions are added with documented justification (`[UnconditionalSuppressMessage]` or `[DynamicallyAccessedMembers]` annotations on the source).
- The check runs on `pr.yaml` so AOT regressions block before merge.

## Notes

- Tier-1 follow-up. Without this, consumers find out at publish time that the library doesn't work in their AOT-published app.
- Originated as part of the IAsyncEnumerable-Extensions thorough-review pass; tracked fleet-wide.
- Board: https://github.com/users/Chris-Wolfgang/projects/6

### LINKED ISSUE #92: [Maintenance] testing: Globalization / CultureInfo invariance (tr-TR, de-DE, zh-CN, ar-SA, ja-JP matrix)
## Scope

The PR test matrix runs only `en-US`. The Turkish dotted-I, German decimal-comma, Chinese collation, and Arabic right-to-left digit-shaping classes of bugs surface only under non-`en-US` cultures. A library that handles strings, dates, or numbers needs explicit verification under hostile cultures.

## Acceptance criteria

- A `tests/.../Globalization/` test set (or `[Theory]` with `CultureInfo` rows on existing tests) runs the suite under: `en-US`, `tr-TR` (dotted-I), `de-DE` (decimal comma), `zh-CN` (collation + simplified Chinese formatting), `ar-SA` (RTL + Hindi-Arabic digit shapes), `ja-JP` (full-width digits).
- `CultureInfo.CurrentCulture` AND `CultureInfo.CurrentUICulture` are both swapped per test.
- `Thread.CurrentThread.CurrentCulture` is restored after each test (xunit fixture).
- A documented allowlist of "intentionally culture-sensitive" public methods exists; everything else is culture-invariant by contract and asserted as such.

## Notes

- Tier-1 follow-up. Especially load-bearing for: DateTime-Extensions (date formatting), String-Extensions (case handling + collation), Try-Pattern (error message formatting), ETL-* (parsing numeric fields).
- Originated as part of the IAsyncEnumerable-Extensions thorough-review pass; tracked fleet-wide.
- Board: https://github.com/users/Chris-Wolfgang/projects/6

### LINKED ISSUE #93: [Maintenance] CI/CD: Build reproducibility verification (deterministic builds across runners)
## Scope

`Directory.Build.props` sets `<ContinuousIntegrationBuild>` and `<Deterministic>true</Deterministic>` — both INPUTS to a reproducible build. Reproducibility is not currently *verified*: building the same commit on two different runners (or after a clean / cache-reset / different OS) should produce byte-identical `.dll` / `.pdb` / `.nupkg` outputs. We don't check.

The deferred work is a build-reproducibility verification workflow that builds the same commit twice (different runners, different OS where possible), and `sha256sum`s the outputs.

## Acceptance criteria

- A `reproducible-build.yaml` workflow builds the current commit on `ubuntu-latest` and `windows-latest`, captures `sha256sum` of every `.dll` / `.pdb` / `.nupkg` produced.
- The two output sets must match byte-for-byte (after handling platform-specific path embeddings — `<PathMap>` MSBuild property + `ContinuousIntegrationBuild` should make this work).
- Any divergence fails the workflow with a diff of which symbols differ.
- A `REPRODUCIBLE-BUILD.md` documents the reproducibility guarantee + how a third party can verify.

## Notes

- Tier-1 follow-up. Required for supply-chain integrity claims (related to the SBOM / SLSA issue from the previous round).
- Reproducibility differs from determinism: deterministic = same inputs produce same outputs; reproducible = same inputs produce same outputs *across different environments*.
- Originated as part of the IAsyncEnumerable-Extensions thorough-review pass; tracked fleet-wide.
- Board: https://github.com/users/Chris-Wolfgang/projects/6

### LINKED ISSUE #101: [Maintenance] testing: Per-PR perf regression detection (BDN delta vs base branch)
## Scope

#184 / P2 (benchmark gh-pages chart) graphs the BDN trend on every push to main — a *backward-looking* signal that surfaces regressions after they merge. The deferred work is the *forward-looking* signal: compare the PR's BDN run against its own base branch and post a delta-table comment on the PR.

A PR that doubles allocations on a hot path should be caught before merge, not after.

## Acceptance criteria

- A `pr-benchmarks.yaml` workflow runs the benchmark suite on the PR's HEAD and on `base...HEAD`'s parent, captures both BDN JSON outputs, computes the per-benchmark delta (time / allocations), and comments on the PR with a markdown table.
- The comment is *replaced* on each push to the PR (not appended) so the latest delta is always at the top.
- The workflow fails the PR if any delta exceeds a configurable threshold (e.g. 20% slower or 50% more allocations) AND the PR didn't tag itself `perf-impact-acknowledged`.
- The workflow only runs when src/ or benchmarks/ files change (path filter).

## Notes

- Tier-1 follow-up. Distinct from #184 (P2 — graphs trend); this is per-PR gating.
- Originated as part of the IAsyncEnumerable-Extensions thorough-review pass; tracked fleet-wide.
- Board: https://github.com/users/Chris-Wolfgang/projects/6

### LINKED ISSUE #102: [Maintenance] security: Reproducible-build third-party verification (consumer side)
## Scope

The reproducible-build verification issue (round 3) tracks proving OUR build is reproducible. The deferred work is the consumer-side flip: documenting how a third party verifies it independently — clones the source at the published tag, builds in their own environment, compares `sha256sum`s against the published NuGet package.

Without published instructions, "our builds are reproducible" is an unverifiable claim. The Reproducible Builds project ([reproducible-builds.org](https://reproducible-builds.org/)) has conventions for this; [vouchsafe.io](https://vouchsafe.io/) and similar can publish third-party verification attestations.

## Acceptance criteria

- A `docs/REPRODUCIBLE-BUILD.md` documents: which exact tooling versions to use, the build command, what `sha256sum`s to expect for each release artifact, and how to file a discrepancy issue.
- Per-release: the `release.yaml` publishes a `reproducible-build-manifest.json` with expected hashes as a release artifact.
- A documented procedure for a third party to upload a verification attestation (Reproducible Builds project conventions or vouchsafe.io).
- The README links the documented procedure under a "Verify the build" section.

## Notes

- Tier-1 follow-up. Pairs with the round 3 reproducible-build verification issue. Together they close the supply-chain integrity loop.
- Originated as part of the IAsyncEnumerable-Extensions thorough-review pass; tracked fleet-wide.
- Board: https://github.com/users/Chris-Wolfgang/projects/6

### PR DESCRIPTION
**Pull request by Chris-Wolfgang**
Release **0.4.0** — `vNext → main`.

## Headline
- **Observability pipeline operators** `Tap`, `Log`, `Throttle` (+ `ThrottleTransformer<T>`) on `IEtlPipeline<T>` (#174) — the one new library surface, layered over `Through`, no new dependencies.

## Also includes (thorough-review hardening — all CI/test/docs)
Reproducible-build verification (#93) + consumer-side + manifest (#102), per-PR BDN delta gate (#101), allocation-budget (#94), cross-platform/multi-arch differential (#86), sustained-load GC profiling (#89), shadow testing (#77), Coyote concurrency (#84), SLSA provenance + SBOM + gated signing (#85), Semgrep SAST (#78), Native-AOT/trim smoke (#90), property-based fuzzing (#76), snapshot testing (#87), globalization matrix (#92), PackageValidation ABI gate (#83).

## Release mechanics
- Version `0.4.0`, AssemblyVersion `0.4.0.0`; PublicAPI Unshipped→Shipped; CHANGELOG `[0.4.0] - 2026-08-09`.
- **Stays on Abstractions 0.20** — the 0.22 bump (#195) is deliberately excluded and remains open for a later release.
- Protected workflow files were split into #204 (already merged), so this PR merges **without** admin-bypass. After merge, tag `v0.4.0`.

## Closes
Closes #174
Closes #76
Closes #77
Closes #78
Closes #83
Closes #84
Closes #85
Closes #86
Closes #87
Closes #89
Closes #90
Closes #92
Closes #93
Closes #101
Closes #102

Remaining open (not delivered in 0.4.0): #197 (0.22 ErrorPolicy), #194 (signing cert), #193 (cross-OS repro hard-fail), #133 (InspectCode ruleset).

### CODE REVIEW

File: .github/workflows/release.yaml [hunk 1]
```diff
@@ -757,8 +795,19 @@ jobs:
...
+      attestations: write    # Required to record the SLSA provenance attestation
     steps:
+      # Checkout first: actions/checkout cleans the workspace by default, so it must
+      # run before the artifacts are downloaded. Provides scripts/generate-repro-manifest.py
+      # at the released tag.
+      - name: Checkout code at release tag
+        uses: actions/checkout@v7
+        with:
+          ref: ${{ github.event.release.tag_name }}
+          persist-credentials: false
```
[the remark this answers is not in the document]
Comment (Chris-Wolfgang) [reply .github/workflows/release.yaml:842]: Addressed in 2a4076e — all four `actions/checkout` references in release.yaml are now pinned to the immutable commit `3d3c42e5aac5ba805825da76410c181273ba90b1 # v7` (the repo-canonical v7 SHA), matching the rest of the workflows. This comment predates that fix.

File: src/Wolfgang.Etl.Transformers/ThrottleTransformer.cs [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+                if (wait > TimeSpan.Zero)
+                {
+                    await _delayAsync(wait, token).ConfigureAwait(continueOnCapturedContext: false);
+                }
+            }
+
+            yield return item;
+
+            hasPrevious = true;
+            previousTimestamp = _getTimestamp();
```
[the remark this answers is not in the document]
Comment (Chris-Wolfgang) [reply src/Wolfgang.Etl.Transformers/ThrottleTransformer.cs]: Fixed. `previousTimestamp` is now recorded at delivery (immediately before `yield return`) instead of after, so `elapsed` spans the consumer's processing time and a consumer already slower than minInterval gets no extra delay — successive items stay >= minInterval apart, adaptively. Added a test whose fake clock advances only during consumer processing (not per read); it fails on the old placement and passes now. 51468cf.

File: tests/Wolfgang.Etl.Transformers.Tests.Concurrency/Scenarios.cs [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+
+    private static async IAsyncEnumerable<int> RangeAsync(int count)
+    {
+        for (var i = 0; i < count; i++)
+        {
+            yield return i;
+        }
+
+        await Task.CompletedTask.ConfigureAwait(false);
```
[the remark this answers is not in the document]
Comment (Chris-Wolfgang) [reply tests/Wolfgang.Etl.Transformers.Tests.Concurrency/Scenarios.cs]: Fixed. RangeAsync now `await Task.Yield()`s per item (replacing the synchronous yields + trailing completed-task await), so MoveNextAsync completes asynchronously — the two consumers in TwoConcurrentEnumerations actually interleave and Coyote can schedule between yields. Concurrency stress 3/3 still green. 51468cf.
